indra/db/util.py

```diff
@@ -472,6 +471,241 @@ def make_stmts_from_db_list(db_stmt_objs):
     return stmts_from_json(stmt_json_list)
 
 
+def get_prefered_tcid(info_tuples, priorities=None, always_include=None):
+    """Filter a list of tcids by prefered source.
+
+    Parameters
+    ----------
+    info_tuples : list[tuple]
+        A list of tuples, each of which containing a group_id, item_id,
+        and quality_id string
+    priorities : list[str]
+        A list of the sources in order of priority, from best to worst.
+    always_includ : list[str]
+        A list of the sources to always include in the final set of tcids.
+    """
+    def is_better(new, old):
+        if new in priorities and old in priorities:
+            return priorities.index(new) < priorities.index(old)
+        return False
+    tcids = set()
+    tr_best = {}
+    for trid, tcid, src in info_tuples:
+        if trid not in tr_best.keys() or is_better(src, tr_best[trid][0]):
+            tr_best[trid] = (src, tcid)
+        if always_include is not None and src in always_include:
+            tcids.add(tcid)
+    tcids |= {tcid for _, tcid in tr_best.values()}
+    return tcids
+
+
+class NestedDict(object):
```

Comment:
I think a couple of unit tests would be nice to make sure this works as intended! I still wonder if something like `defaultdict(defaultdict)` could achieve a similar structure.

Reply:
I agree. I'm actually working on those now. The `get`, `get_path`, `gets`, and `get_paths features` would not be available from the nearest equivalent using `defaultdict`. I'm still not completely convinced that the added features are worth adding the features, though. They could certainly make life easier and more convenient in some cases, but I'm not sure this is that case.